### PR TITLE
provider/aws: Opsworks instance

### PR DESCRIPTION
**Pull request by mrjefftang**
This is a work in progress to support issue #3459.

There's still a lot of work to be done.

TODO:
- [x] CRUD functionality
- [x] Code cleanup
- [x] Documentation

### CODE REVIEW

**Comment by mrjefftang** [comment]
P.S.  Any comments / suggestions would be greatly appreciated.

**Comment by jmcclell** [comment]
@mrjefftang I actually have this implemented already. I haven't created a PR yet because I'm still testing and also trying to add opsworks app, the ability to register existing ec2 instances via an opsworks_instance_registration resource, and trying to figure out the best way to do instance scheduling for time based instances. You can see the current stuff I have pushed at https://github.com/jmcclell/terraform. Currently it's in the mater branch.

I'm very new to Go so if you have any feedback I'd appreciate it.

**Comment by mrjefftang** [comment]
@jmcclell Looks like we've came to mostly the same code with the exception that you've added some additional validation and I've handled root block device definitions.

**Comment by jmcclell** [comment]
Root block devices aren't part of the opsworks instance level, as far as I can tell?

**Comment by mrjefftang** [comment]
You can actually define the root block device via the API.  I've tested it with the code on this pull request.

I've also incorporated your validation checks into my branch via the ValidateFunc specification on the schema.

**Comment by jmcclell** [comment]
Aye. It looks like you've done some other neat stuff, as well. Elastic IP/Vol options, for example. Some of my validation checks are actually probably not necessary since the AWS struct itself will yell - I didn't realize this at first (again, new to Go :) )

Curious - what is the advantage of pulling back items such as CreateTime, etc? I wasn't sure if it made sense to do so or not since it isn't part of managing the resource. Since I'm very new to all of this perhaps I am missing something?

**Comment by mrjefftang** [comment]
I figure validation should be front loaded as possible.

I pulled back all of the extra attributes just out of completeness, I was just going down the list from the structure definition in the AWS SDK.

I haven't thought about the instance state yet, not sure what the different use cases will be.  It makes sense to have a "desired state" and react to that.

**Comment by mrjefftang** [comment]
This PR is completed.

File: builtin/providers/aws/resource_aws_opsworks_instance.go [hunk 1]
```diff
@@ -0,0 +1,1001 @@
...
+func validateVirtualizationType(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "paravirtual" && value != "hvm" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"paravirtual\" or \"hvm\"", k))
+	}
+	return
+}
+
+func resourceAwsOpsworksInstanceValidate(d *schema.ResourceData) error {
```
Comment (jmcclell) [inline builtin/providers/aws/resource_aws_opsworks_instance.go]: Are you certain this is the case? I don't see this in the documentation, but haven't had a chance to test myself. Seems odd that this would be restricted since you can add block devices to any other ec2 instance regardless of AMI.

Comment (mrjefftang) [inline builtin/providers/aws/resource_aws_opsworks_instance.go]: It errors out when attempting to add a block device mapping when using a
custom AMI.

I believe the only way to do it is by specifying the block device mapping
when creating your custom AMI.

**Comment by cmcarthur** [comment]
:+1:

**Comment by jmcclell** [comment]
Cool deal! I'll finish up opsworks application and submit that, as well.

Any ideas on run state?

**Comment by mrjefftang** [comment]
@jmcclell 'state' is provided. If set to "running" it will start the instance, if set to "stopped" it will stop it.

**Comment by cmcarthur** [comment]
:+1:

File: builtin/providers/aws/provider.go [hunk 2]
```diff
@@ -240,6 +240,7 @@ func Provider() terraform.ResourceProvider {
 			"aws_opsworks_mysql_layer":             resourceAwsOpsworksMysqlLayer(),
 			"aws_opsworks_ganglia_layer":           resourceAwsOpsworksGangliaLayer(),
 			"aws_opsworks_custom_layer":            resourceAwsOpsworksCustomLayer(),
+			"aws_opsworks_instance":		resourceAwsOpsworksInstance(),
```
Comment (beedub) [inline builtin/providers/aws/provider.go]: is this causing the build failure?

File: builtin/providers/aws/resource_aws_opsworks_instance.go [hunk 3]
```diff
@@ -0,0 +1,998 @@
...
+	if err != nil {
+		return err
+	}
+
+	if wait {
+		log.Printf("[DEBUG] Waiting for instance (%s) to become running", instanceId)
+
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"requested", "pending", "booting", "running_setup"},
+			Target:     "online",
```
Comment (beedub) [inline builtin/providers/aws/resource_aws_opsworks_instance.go]: i think the build failure indicates this needs to be `[]string{"online"}`?

File: builtin/providers/aws/resource_aws_opsworks_instance.go [hunk 4]
```diff
@@ -0,0 +1,998 @@
...
+	if err != nil {
+		return err
+	}
+
+	if wait {
+		log.Printf("[DEBUG] Waiting for instance (%s) to become stopped", instanceId)
+
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"stopping", "terminating", "shutting_down", "terminated"},
+			Target:     "stopped",
```
Comment (beedub) [inline builtin/providers/aws/resource_aws_opsworks_instance.go]: same with this `[]string{"stopped"}`

**Comment by mrjefftang** [comment]
Thanks @beedub .  

It looks like there were a few changes in Terraform after this PR was submitted.  

I had been using this code successfully until now, I'll give it a quick look and try to get it working again.

**Comment by beedub** [comment]
np, sorry if these comments came off as pushy. i'm personally very interested in getting these changes merged, so just trying to help :)

**Comment by mrjefftang** [comment]
No worries.  I didn't see them as pushy.  I completely understand wanting to be able to use the functionality.

I lost interest in getting these merged a few months ago since there was no response from the maintainers.  I was using a local build of terraform with my changes so I could use them internally.

**Comment by mrjefftang** [comment]
Added acceptance testing which is the final requirement

**Comment by mrjefftang** [comment]
@apparentlymart @radeksimko Any updates on getting this merged?  I just rebased it again to fix a merge conflict as this PR.

**Comment by jonathanq** [comment]
I am going to start looking at terraform and I use OpsWorks heavily, so being able to define the instances would be nice. Hope this gets merged soon!

**Comment by mrjefftang** [comment]
@phinze @jen20 What else is needed to get this merged?

**Comment by phinze** [comment]
Hi @mrjefftang - sorry about the delay here - great work on this so far! I have a few minor comments that I'll make in line, but this is really close to being ready to ship!

File: builtin/providers/aws/resource_aws_opsworks_instance.go [hunk 5]
```diff
@@ -0,0 +1,998 @@
...
+			if awserr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+		}
+		return err
+	}
+
+	instance := resp.Instances[0]
+	instanceId := *instance.InstanceId
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_opsworks_instance.go]: We've learned in our time with the AWS API that all sorts of different shapes of responses are possible! :grinning: 

So we tend to be extra defensive in handling types out of the SDK - I'd double check both the length of `resp.Instances` and that the `InstanceId` pointer is non-null. (For the rest of the `d.Set` calls below the null check is handled in the implementation so we're good there.)

File: website/source/docs/providers/aws/r/opsworks_instance.html.markdown [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+```
+resource "aws_opsworks_instance" "my-instance" {
+  stack_id = "${aws_opsworks_stack.my-stack.id}"
+  layer_ids = [
+    "${aws_opsworks_custom_layer.my-layer.id}",
+  ]
+  instance_type = "t2.micro"
+  os = "Amazon Linux 2015.09"
+  state = "stopped"
```
Comment (phinze) [inline website/source/docs/providers/aws/r/opsworks_instance.html.markdown]: Style nit - can you format this like `hclfmt` would? (Basically align equals for adjacent attributes.)

File: builtin/providers/aws/resource_aws_opsworks_instance_test.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+					),
+					resource.TestCheckResourceAttr(
+						"aws_opsworks_instance.tf-acc", "os", "Amazon Linux 2014.09", // inherited from opsworks_stack_test
+					),
+					resource.TestCheckResourceAttr(
+						"aws_opsworks_instance.tf-acc", "root_device_type", "ebs", // inherited from opsworks_stack_test
+					),
+					resource.TestCheckResourceAttr(
+						"aws_opsworks_instance.tf-acc", "availability_zone", "us-west-2a", // inherited from opsworks_stack_test
+					),
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_opsworks_instance_test.go]: These tests are really looking good overall! One more thing they need is a `TestCheckExists` and `TestCheckAttributes` function - these are the two test checks that actually interact with the AWS APIs to ensure that everything looks correct.

If you take a look at some of the sibling tests you should be able to follow the pattern - `TestCheckExists` verifies the resource can be looked up and sets a pointer to the resource for `TestCheckAttributes` to analyze.

Comment (mrjefftang) [inline builtin/providers/aws/resource_aws_opsworks_instance_test.go]: Can you point to an example?  I don't see any instances of `TestCheckExists` nor `TestCheckAttributes`

Comment (mrjefftang) [inline builtin/providers/aws/resource_aws_opsworks_instance_test.go]: After waking up some more and with some more coffee, it makes sense now.  I'm working on adding the tests.

**Comment by u2mejc** [comment]
@mrjefftang Awesome work, I'm looking forward to having this merged as well. :shipit:

**Comment by stack72** [comment]
Hi @mrjefftang 

thanks for fixing up the comments that @phinze made, I tried to have another look at this (by running the tests) and get the following:

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSOpsworksInstance'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSOpsworksInstance -timeout 120m
=== RUN   TestAccAWSOpsworksInstance
--- FAIL: TestAccAWSOpsworksInstance (13.67s)
	testing.go:154: Step 0 error: Error applying: 1 error(s) occurred:

		* aws_opsworks_stack.tf-acc: ValidationException: Vpc Id: Failed to check with EC2 - please validate IAM role permissions
			status code: 400, request id: 1965ec6d-fadb-11e5-8764-23fe7fda5cec
	testing.go:175: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Fall through error on OpsWorks instance test

		State: <no state>
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/aws	13.689s
make: *** [testacc] Error 1
```

I know that at the top of the file you have the following:

```
// These tests assume the existence of predefined Opsworks IAM roles named `aws-opsworks-ec2-role`
// and `aws-opsworks-service-role`.
```

but we cannot guarantee that these roles exist. What is needed here to include these roles as part of the test?

Thanks

Paul

**Comment by mrjefftang** [comment]
Hi @stack72 ,

That comment is out of date as it was copied from `resource_aws_opsworks_custom_layer_test.go`.  It no longer relies on the predefined opsworks roles.  I'll remove it.

The instance test relies on `testAccAwsOpsworksStackConfigVpcCreate` in `resource_aws_opsworks_custom_layer_test.go` which actually defines custom IAM roles as of #4009.

What you have run into is an edge case I've experienced where the IAM role is created but is not available for use.  It's a race condition between Terraform attempting to create an Opsworks stack using the newly created IAM role and AWS propagating the existence of the role.

In using my code for some test environments, just rerunning Terraform after a few minutes (once the IAM roles were created in the first run) works just fine.

```
 09:09:47  jeff@apollo  ...github.com/hashicorp/terraform  ♦️ system  🐍 system   opsworks_instance ✘ ✹   $  make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSOpsworksInstance'
==> Checking that code complies with gofmt requirements...
/Users/jeff/Projects/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSOpsworksInstance -timeout 120m
=== RUN   TestAccAWSOpsworksInstance
--- FAIL: TestAccAWSOpsworksInstance (15.73s)
	testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

		* aws_opsworks_stack.tf-acc: ValidationException: Vpc Id: Failed to check with EC2 - please validate IAM role permissions
			status code: 400, request id: 13485edb-fb30-11e5-85e6-a119075d1df9
	testing.go:169: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Fall through error on OpsWorks instance test

		State: <no state>
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/aws	15.760s
make: *** [testacc] Error 1

 09:12:48  ✘ 2  jeff@apollo  ...github.com/hashicorp/terraform  ♦️ system  🐍 system   opsworks_instance ✘ ✹   $ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSOpsworksInstance'
==> Checking that code complies with gofmt requirements...
/Users/jeff/Projects/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSOpsworksInstance -timeout 120m
`=== RUN   TestAccAWSOpsworksInstance
--- PASS: TestAccAWSOpsworksInstance (38.20s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	38.224s
```

**Comment by mrjefftang** [comment]
On subsequent runs with the IAM roles destroyed and recreated it also works fine.  Looking through the AWS webconsole, I can find no dangling resources.

**Comment by u2mejc** [comment]
@mrjefftang Would it possible to put a waiter in there for the target resource to become available and avoid the race condition? I think this would be ready to merge then after the tests are stabilized, would you agree @stack72?

**Comment by mrjefftang** [comment]
@u2mejc Looks like it was attempted by @stack72 in #4235 but not accepted.

Also @phinze had an attempt for retryable errors in #4844.

The alternative is to submit a separate PR as a bugfix for Opsworks stack with something similar to https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_instance.go#L363

The problem is I can't reproduce this issue most of the time.  Also this error is slightly different than what I've experienced previously in #2162.  Before the actual IAM profile didn't exist.  Now it just looks like the permissions for the IAM aren't propagated yet.

**Comment by u2mejc** [comment]
@mrjefftang good call adding to the waiter to prevent errors while it propagates. @stack72 , https://github.com/hashicorp/terraform/issues/6049 should resolve the testing issue. Are there other issues you'd like to see addressed?

**Comment by apparentlymart** [comment]
Thanks for this, @mrjefftang! And again, sorry for the delay in getting these opsworks features merged.

I added a small followup fix da839e765c4f6cd6d65a6dd50614ccb6fbe0c95f for a slightly-broken config example in the docs.
